### ISSUE TITLE
[Bug]: cheap-[module]-source-map not work with react compiler

### ISSUE THREAD

**Issue by xfsnowind** [web-infra-dev/rspack#6793]
### System Info

System:
    OS: macOS 14.5
    CPU: (10) arm64 Apple M1 Pro
    Memory: 323.69 MB / 16.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 18.18.1 - ~/.nvm/versions/node/v18.18.1/bin/node
    npm: 9.8.1 - ~/.nvm/versions/node/v18.18.1/bin/npm
    pnpm: 9.3.0 - ~/.nvm/versions/node/v18.18.1/bin/pnpm
    bun: 1.0.14 - ~/.bun/bin/bun
  Browsers:
    Brave Browser: 123.1.64.116
    Chrome: 119.0.6045.105
    Edge: 119.0.2151.97
    Safari: 17.5
  npmPackages:
    @rspack/cli: 0.7.2 => 0.7.2
    @rspack/core: 0.7.2 => 0.7.2
    @rspack/plugin-react-refresh: 0.7.2 => 0.7.2

### Details

Now we are using react-compiler for our dev branch code with version 0.6.5, the source map can be generated from react-compiler with `devtool: 'cheap-module-source-map'`

But it doesn't work with version 0.7. The generated source maps are not compiled by compiler and breakpoint won't work.

If I change the `devtool` to `'source-map'`, breakpoint would work, but the generated source map is not compiled by compiler.

Here is the part of rspack config:
```js
  devtool: 'cheap-module-source-map',
  module: {
    parser: {
      //...
    },
    rules: [
      ...getModuleJSAndTSRules({ useReactRefresh }),
    ...,
   ],
...
```

and the method `getModuleJSAndTSRules`:
```js
getModuleJSAndTSRules: ({ useReactRefresh }) => {
    if (typeof useReactRefresh !== 'boolean') {
      throw new Error('[getModuleJSAndTSRules] - Please provide valid parameters')
    }

    const swcLoaderConfig = {
      loader: 'builtin:swc-loader',
      /** @type {import('@rspack/core').SwcLoaderOptions} */
      options: {
        sourceMap: true,
        env: {
          targets: 'defaults and fully supports es6-module',
        },
        jsc: {
          parser: {
            syntax: 'typescript',
            tsx: true,
            dynamicImport: true,
          },
          transform: {
            react: {
              runtime: 'automatic',
              refresh: useReactRefresh,
              development: useReactRefresh,
            },
          },
        },
      },
    }

    return [
      {
        test: /\.(js|ts|jsx)$/,
        include: resolve(__dirname, 'src'),
        /** @type {import('@rspack/core').RuleSetUse} */
        use: [swcLoaderConfig],
      },
      {
        test: /\.tsx$/,
        include: resolve(__dirname, 'src'),
        /** @type {import('@rspack/core').RuleSetUse} */
        use: (info) => {
          // Configuration based on https://github.com/rspack-contrib/rspack-examples/tree/main/rspack/react-compiler-babel-ts

          // Do not run react compiler on these places since they tend to often violate the react rules at the moment and/or are deprecated.
          if (
            !info.realResource.includes('/modules/privacy') &&
            !info.realResource.includes('/modules/delivery') &&
            !info.realResource.includes('/modules/mifleet')
          ) {
            return [
              swcLoaderConfig,
              {
                loader: 'babel-loader',
                options: {
                  plugins: [
                    [
                      'babel-plugin-react-compiler',
                      {
                        runtimeModule: 'react-compiler-runtime',
                        // panicThreshold: 'critical_errors',
                        // environment: {
                        //   enableResetCacheOnSourceFileChanges: true,
                        // },
                      },
                    ], // must run first!
                    ['@babel/plugin-syntax-typescript', { isTSX: true }],
                  ],
                },
              },
            ]
          }

          return [swcLoaderConfig]
        },
      },
    ]
  }
```

### Reproduce link

[react compiler rspack ts example](https://github.com/xfsnowind/react-compiler-rspack-ts-example)

### Reproduce Steps

1. git clone [this repo](https://github.com/xfsnowind/react-compiler-rspack-ts-example)
2. open the file `package.json` and update the rspack libs' version to `0.6.5`
<img width="647" alt="image" src="https://github.com/web-infra-dev/rspack/assets/2863328/09db89ac-b49c-45c2-902d-e15138e8d5d6">

3. run `pnpm i` and `pnpm run dev`
4. open the page `localhost:8080` and the inspector `source` tab
<img width="1163" alt="image" src="https://github.com/web-infra-dev/rspack/assets/2863328/7676321d-84a6-464c-9185-4cf787c39137">

5. in the source map file, the react codes are compiled by compiler and set the breakpoint to the `setData("data")` line. Every time we click on the `click` button, breakpoint would work
6. stop the terminal and open the `package.json`, revert the rspack libs' version back to `0.7.2`
7. do the same thing, the source map files are not compiled and break point not work:
<img width="1142" alt="image" src="https://github.com/web-infra-dev/rspack/assets/2863328/8ced48e7-28d9-481e-b363-00cdb46de9e1">

**Comment by h-a-n-a** [comment]
I reproduced this locally and it turned out the `cheap` source map is not correct on both v0.6.5 and the latest version. Thank you for the heads-up.

**Comment by pushkargupta9807** [comment]
@h-a-n-a @xfsnowind 
I am also observing the same issue with "cheap-module-source-map". 

I also reported a different issue below with source-maps and rspack, hopefully with the local repro for this issue you can also repro  https://github.com/web-infra-dev/rspack/issues/6714

Basically for webpack we are able to successfully use "eval-cheap-module-source-map".  But with rspack, with this option the devTools in edge browser do not have the folder structure maintained and everything seems to load with giant relative paths. The bug has images . Eventually the devTool simply freezes and then crashes.

**Comment by h-a-n-a** [comment]
Ok, here's what I discovered locally. The problem is in the `rspack_sources`, which was leveraged in the `rspack_plugin_devtool` here:
https://github.com/web-infra-dev/rspack/blob/bcb70bebf7c11b778c50f29b1417fe42fcae05a1/crates/rspack_plugin_devtool/src/lib.rs#L231-L232
Here we mapped the sources into code and source maps. `columns` was determined upon the `devtool` you passed in.
It's `false` if it's a cheap source map. This is where everything goes wrong.

A simplified version to reproduce it locally:
1. Set `devtool` as `cheap` source map, it can be `cheap-source-map` for example
3. Create two files, one imports the other one
4. The final generated assets showed the source map is wrong
